### ISSUE TITLE
PDBParser support for pdb.gz files

### ISSUE THREAD

**Issue by duerrsimon** [biopython/biopython#3498]
### Setup

I am reporting a problem with Biopython version 1.78, Python version 3.7.8, and Linux-4.19.0-13-amd64-x86_64-with-debian-10.  as follows:

### Expected behaviour

Bio.PDB.PDBParser should read also *.pdb.gz files

### Actual behaviour

Parser does not read pdb.gz files. 

### Proposed fix

Create a  PDBGZParser or augment the PDBParser with a selection based on file ending. 
An extra step is necessary to decode the uncompressed string to utf8.  I am happy to contribute this via Pull request. Let me know if a new `PDBGZParser` class or an update to PDBParser is preferred.

Users might prefer to not uncompress pdb.gz files on disk if they process many of those.

as_handle function
```
import gzip

@contextlib.contextmanager
def as_handle_gzip(handleish, mode="r", **kwargs):
     try:
        with gzip.open(handleish, mode, **kwargs) as fp:
            yield fp
    except TypeError:
        yield handleish
```

changes to `get_structure` function 
```
            with as_handle_gzip(file) as handle:
                lines = handle.readlines()
                # decode lines
                lines= [line.decode('utf-8') for line in lines]
                if not lines:
                    raise ValueError("Empty file.")
                self._parse(lines)
```

**Comment by JoaoRodrigues** [comment]
Hi @duerrsimon 

Thanks for raising this issue - it's definitely a good improvement to include. Besides `PDBParser`, `MMCIFParser` would also benefit from such a function. Would you be willing to provide a patch for both of them?

Thanks!

**Comment by duerrsimon** [comment]
Sure. Happy to contribute a pull request for both.

**Comment by JoaoRodrigues** [comment]
Awesome - let me know if you need any help!

**Comment by mdehoon** [comment]
It may be better to add a `read` function to `Bio.PDB` (instead of the current design relying on a `PDBParser` class) that can take both a filename and a file handle. Then you can use `gzip.open` to open gzipped files and pass the result to `read`. This will allow you to also handle e.g. bzipped files, as well as remote PDB files accessed through `urlopen`. For this reason other modules (for example `SeqIO`, `AlignIO`, `Entrez`, `SwissProt`, `Sequencing`) in Biopython use a `read` (and `parse`) function.

**Comment by JoaoRodrigues** [comment]
Indeed, I thought of the same thing, living in `Bio.PDB.__init__.py`.

**Comment by duerrsimon** [comment]
I have added my solution now as a pull request. I feel that refactoring the Bio.PDB code to be consistent with other functions in biopython is probably something that a more experienced maintainer should do. 

@mdehoon PDBParser and MMCIFParser already accept filenames or file handles.

**Comment by mdehoon** [comment]
> @mdehoon PDBParser and MMCIFParser already accept filenames or file handles.

Then instead of this PR users can just pass the result of `gzip.open` to `PDBParser().get_structure`.

>  I feel that refactoring the Bio.PDB code to be consistent with other functions in biopython is probably something that a more experienced maintainer should do.

Adding a `read` function to `Bio.PDB` is very easy and will give you something much more user-friendly than the current `PDBParser().get_structure`. It doesn't require refactoring code.

**Comment by duerrsimon** [comment]
> > @mdehoon PDBParser and MMCIFParser already accept filenames or file handles.
> 
> Then instead of this PR users can just pass the result of `gzip.open` to `PDBParser().get_structure`.
> 

Nope not that easy unfortunately. The output from gzip.open is still a byte string which needs to be decoded in the Parser.

**Comment by mdehoon** [comment]
@duerrsimon 
> Nope not that easy unfortunately. The output from gzip.open is still a byte string which needs to be decoded in the Parser.

You need to open the file in text mode, as in `gzip.open("myfile.pdb.gz", "rt")`. For some reason, `gzip.open` opens the file in binary mode by default, whereas `open` on uncompressed files opens the file in text mode.

**Comment by duerrsimon** [comment]
Thanks for the pointer! I have modified my code and removed the decoding in the pull request. I still think it might be more user friendly if gz files can also be read from a filename. 
Especially since the default mode of `gzip.open` is binary which will make the code fail if the text mode option is not set if a filehandle from gzip.open is passed.

**Comment by mdehoon** [comment]
The problem that I am seeing here is that having to create a parser first with `PDBParser` and then parsing the file by calling `get_structure` is not user-friendly and inconsistent with other modules in Biopython. This should be changed to a `read` function in `Bio.PDB` at some point; the sooner the better.

This PR does not fix that but adds automatic decompression of gzipped files to the `get_structure` function. We cannot evaluate now whether the proposed solution is the right way to handle automatic decompression of gzipped files once we move to `Bio.PDB.read`. So that has to happen first.

Also, keep in mind that it is very easy to add code to Biopython, but very difficult to remove it later. It requires obsolete warnings, deprecation warnings, and a year waiting time before we can actually remove the code. Seriously, that stuff is a headache and delays everything. Then I don't want to add code now if we already know that we have to rethink it later.

**Comment by duerrsimon** [comment]
Yeah I get that. But wouldn't that also mean that other functions in `Bio.PDB` should move so that everything is consistent again? 

An example below with the current approach
```
from Bio.PDB import *
pdbl = PDBList()
pdbl.retrieve_pdb_file("1FAT")
parser=PDBParser()
structure=parser.get_structure("example", "1FAT.pdb")
```
and a more readable approach as discussed above consistent with the change to the read function. 
```
from Bio import PDB
pdbl = PDB.retrieve_pdb_file("1FAT")
structure = PDB.read("example", "1FAT.pdb") 
```
If the read function is added but the other stuff is kept untouched it would be: 
```
from Bio import PDB
from Bio.PDB import *
pdbl = PDBList()
pdbl.retrieve_pdb_file("1FAT")
structure=PDB.read("example", "1FAT.pdb")
```

Another question to me would be whether there should then also be Bio.PDB.parse() function with returns an iterator over the Models in the structure? This would then be similar to `SeqIO` parse

**Comment by mdehoon** [comment]
@duerrsimon  
It's up to you how far you want to go; you don't need to fix everything at the same time. If we all agree that a `PDB.read` function is a good idea, then you may want to leave `PDB.parse` and `PDB.retrieve_pdb_file` to somebody else.

That said, if you can provide a comprehensive solution with `PDB.read`, `PDB.parse`, and `PDB.retrieve_pdb_file`, then so much the better. The interface to `Bio.PDB` is a bit antiquated and it would be good to bring it up to date. I would just suggest to describe your proposal here before coding it up to get feedback from the other developers.

For a `read` function, I guess it makes sense if it takes a filename (or file handle) only, without the `id` argument. I never quite understood why `PDB.PDBParser().get_structure` takes both an `id` and a `file` argument; as far as I know the `id` can be read from the PDB file itself.

**Comment by JoaoRodrigues** [comment]
Hi @duerrsimon,

This is incredible work so far. I can help you with structuring things later this week. I agree 200% with @mdehoon about bundling these two changes (opening gz files and the read() function) together.

The way I would go about this would be to add a function `read()` directly in `Bio.PDB.__init__.py`. I wrote such a function a while back in another library I work on that can serve as inspiration: [link](https://github.com/JoaoRodrigues/interfacea/blob/main/interfacea/io/__init__.py).

Regarding the `id`, it can be left as a keyword argument to `read()` or automatically taken as the root of the filename (ie, name without extension).

**Comment by peterjc** [comment]
I'm thinking this should match the Bio.SeqIO.read(...) style and take a filename/handle and format (PDB etc - probably in lower case for consistency).
